packages/components/src/progressive-disclosure-panel/index.js

```diff
@@ -0,0 +1,140 @@
+	// This panel is only interested in the children to be displayed.
+	const filteredChildren = Array.isArray( children )
+		? children.filter( Boolean )
+		: [];
+
+	// Collect which children have custom values. Used to update menu state to
+	// reflect customization for children that display by default / always show.
+	const customizedChildren = filteredChildren.map( ( child ) =>
+		child.props.hasValue( child.props ) ? child.props.label : undefined
+	);
+
+	// On first render determine initial menu state and which children should
+	// always display by default.
+	useEffect( () => {
+		const items = {};
+		const defaults = {};
+
+		filteredChildren.forEach( ( child ) => {
+			items[ child.props.label ] = child.props.hasValue( child.props );
+			defaults[ child.props.label ] = child.props.isShownByDefault;
+		} );
+
+		setMenuItems( items );
+		setDefaultChildren( defaults );
+	}, [] );
+
+	// As the default children are visible all the time. Reflect their
+	// customizations in the menu items' selected state.
+	useEffect( () => {
+		const menuLabels = Object.keys( menuItems );
+
+		// Skip if no children or menu state not initialized.
+		if ( menuLabels.length === 0 ) {
+			return;
+		}
+
+		const updatedItems = { ...menuItems };
+		menuLabels.forEach( ( label ) => {
+			if ( defaultChildren[ label ] ) {
+				updatedItems[ label ] = customizedChildren.includes( label );
+			}
+		} );
+
+		setMenuItems( updatedItems );
+	}, customizedChildren );
+
+	if ( filteredChildren.length === 0 ) {
+		return null;
+	}
+
+	const getChildByMenuLabel = ( label ) => {
+		return filteredChildren.find(
+			( child ) => child.props.label === label
+		);
+	};
+
+	// Toggles the customized state of the child and its display if it isn't to
+	// be displayed by default. When toggling a child it's callback is executed.
+	const toggleChild = ( label ) => {
+		const wasSelected = menuItems[ label ];
+		const child = getChildByMenuLabel( label );
+		const { onDeselect = noop, onSelect = noop } = child.props;
+
+		if ( wasSelected ) {
+			onDeselect( child.props );
+		} else {
+			onSelect( child.props );
+		}
+
+		setMenuItems( {
+			...menuItems,
+			[ label ]: ! wasSelected,
+		} );
+	};
+
+	// Resets display of children and executes resetAll callback if available.
+	const resetAllChildren = () => {
+		resetAll();
+
+		// Turn off menu items unless they are to display by default.
+		const resetMenuItems = {};
+
+		filteredChildren.forEach( ( child ) => {
+			resetMenuItems[ child.props.label ] = false;
+		} );
+
+		setMenuItems( resetMenuItems );
+	};
+
+	const classes = classnames(
+		'components-progressive-disclosure-panel',
+		className
+	);
+
+	return (
+		<div className={ classes }>
+			<ProgressiveDisclosurePanelTitle
+				menuLabel={ menuLabel }
+				title={ title }
+				menuItems={ menuItems }
+				toggleChild={ toggleChild }
+				resetAll={ resetAllChildren }
+			/>
+			{ filteredChildren.map( ( child ) => {
```

Comment:
Before we merge and publish the component, should we rename it from `ProgressiveDisclosurePanel` to something else? CC: @mtias

Reply:
> Each time the ProgressiveDisclosurePanelTitle renders we generate a new array reference of children even if nothing changed. I think this may cause unexpected rerenders.

This has been fixed up. Thanks!

> Before we merge and publish the component, should we rename it from ProgressiveDisclosurePanel to something else?

The `ProgressiveDisclosurePanel` based names for components are unwieldy and long. The current collapsible panels are simply named `Panel`. I'm definitely open to suggestions on a new, concise name for this panel that still makes its behaviour clear.

Reply:
Material design calls the expand/collapse panel an "ExpansionPanel" (though I think just Panel is fine). But inspired by that, here's a quick brainstorm:

- MorePanel
- ExtensionPanel
- RevealPanel
- ToolsPanel
- VisibilityPanel
- MenuPanel
- ListPanel
- IncrementalPanel
- ControlsPanel
- DropdownPanel
- TogglePanel
- ActionPanel
- DisplayPanel
- RowPanel
- ContextualPanel
- ContextPanel
- CascadePanel

Of those, I like:

- ListPanel
- ToolsPanel
- DropdownPanel

But I'd appreciate additional thoughts!

Reply:
I think Joen's list covers most of the better ones however some other random options could be:

- ProgressivePanel
- DiscoveryPanel
- OptionsPanel
- DynamicPanel
- SelectionPanel

The ListPanel or ToolsPanel options are likely the most concise while still reflecting a little about the panel's behaviour.

Reply:
ToolsPanel I could get behind.